docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,2624 @@
+
+* Seggregate avaialble seeds into `multi-zonal` seeds and `non-multi-zonal` seeds.
+* Assign selection priority based on shoot spec HA configuration for the control plane.
+  * Shoot with `multi-zonal` control plane can only be scheduled on `multi-zonal` seed. Therefore for these shoots the target list of seeds will only be the ones which are `multi-zonal`.
+  * Shoot with `non-multi-zonal` control plane can be scheduled either on `non-multi-zonal` seed or `multi-zonal` seed. In order to find the best fit - first check if there is any available seed in `non-multi-zonal` seeds. If there is one then that is assigned. If there is no available seed in `non-multi-zonal` seeds then look for an available seed in `multi-zonal` seeds.
+
+The above selection algorithm should be applied only after applying the [strategy](https://github.com/gardener/gardener/blob/4e2d28f2af7093eb94819652a6f4709b5fbfaf06/pkg/scheduler/controller/shoot/reconciler.go#L256-L274). This ensures that strategy based selection e.g `region affinity` or `minimum distance` is applied first and then selection of a `single-zone` or `multi-zone` seed cluster is made to optimize the geographic separation between the shoot cluster and the shoot cluster control plane.
+
+**Case #2**
+
+A shoot has a pre-defined non-HA seed. A change has been made to the shoot spec, setting control HA to `multi-zone`. Garden scheduler needs to react to the change in the HA configuration for the shoot control plane.
+
+_Proposed Changes_
+
+* [scheduler manager](https://github.com/gardener/gardener/blob/4dbba56e24b5999fd728cb84b09a4d6d54f64479/pkg/scheduler/controller/shoot/scheduler.go#L57) should remove the predicate `ShootIsUnassigned` and instead define a new predicate which allows a `reconcile.Request` to be created if one of the following is satisified:
+  * Shoot does not have any assigned seed
+  * Shoot spec has changed requesting an uprade from non-HA control plane to a HA control plane. This can be checked by checking the HA configuration in the spec and compare it with the `status` to determine if the upgrade requested is from a non-HA shoot control plane to a HA shoot control plane.
+* [reconciler](https://github.com/gardener/gardener/blob/4dbba56e24b5999fd728cb84b09a4d6d54f64479/pkg/scheduler/controller/shoot/reconciler.go#L52-L99) - function now needs to additionally check if an upgrade request has been made. If it is a valid upgrade request for the shoot control plane then the reconciler should unassign the current seed for this shoot. A new seed appropriate to the HA control plane requirement will be assigned. In this case since there is a change in the seed assignment, `control-plane-migration` flow will kick in and will commence the migration of the control plane from the non-HA seed to a HA seed.
```

Comment:
Uh, do we really want to trigger CPM? Alternative is to decline scheduling/provide error message and ask to recreate cluster from scratch. If that’s rejected/escalated, ops can invoke DR (disadvantage), but we can then tell them that it comes with a significant downtime (that may further motivate them to reconsider). If we want to trigger CPM, we must a.) make the end-user somehow aware that this will come with said significant downtime (only possible in Dashboard) and b.) we must be 100% certain, that CPM is a reliable/safe feature as probably these clusters will be productive that are then rescheduled/DR’ed.

Reply:
Can you explain what you mean by saying "invoke DR"? Is it to copy the etcd data from one cluster to another (created from scratch)?

Reply:
It's doing CPM, changing a shoot's seed. The above paragraph plans to automate that. I am not that sure, we really want that.